### PR TITLE
[Post Processing Serial Number] Introducing Custom Substitution

### PR DESCRIPTION
**Pull request by ruicliu**
For cases such as D9 where bot interprets as "denied" after speech-to-text, we now have the ability for developers to add their own substitutions in a json file directly in the same directory as appsettings.json in the skill project, and the core post processing logic in SerialNumberPattern will try to replace all occurrence with the desired characters (e.g. replace all occurrence of "denied" with "D9"). For more instruction, please see SerialNumber.md.

### CODE REVIEW

File: experimental/SerialNumPostProcessingSkillSample/SpeechSerialNumber/SerialNumberPattern.cs [hunk 1]
```diff
@@ -67,6 +77,8 @@ public class SerialNumberPattern
                 }
             };
 
+        private ConcurrentDictionary<string, ConcurrentDictionary<string, string>> substitutionMapping;
```
Comment (tsuwandy) [inline experimental/SerialNumPostProcessingSkillSample/SpeechSerialNumber/SerialNumberPattern.cs:80]: private static readonly ConcurrentDictionary<string, ConcurrentDictionary<string, string>> SubstitutionMapping;?

Same casing as the other dictionaries?

File: experimental/SerialNumPostProcessingSkillSample/SpeechSerialNumber/SerialNumberPattern.cs [hunk 2]
```diff
@@ -373,6 +423,21 @@ public string[] Inference(string inputString)
                     ambiguousInputIndexes.Add(inputIndex);
                     fixedUpString += '*'; // Mark ambiguity
                 }
+                else if (DetectCustomSubstitutionFixup(InputString, inputIndex) == FixupType.Custom)
```
Comment (tsuwandy) [inline experimental/SerialNumPostProcessingSkillSample/SpeechSerialNumber/SerialNumberPattern.cs:426]: Could you move DetectCustomSubstitutionFixup inside InferMatch?  That way you won't have to deal with patternIndex.

**Review by tsuwandy — CHANGES_REQUESTED** [review]

File: experimental/SerialNumPostProcessingSkillSample/SpeechSerialNumber/SerialNumber.md [hunk 3]
```diff
@@ -97,6 +97,37 @@ Sample usage:
...
+## Custom Substitution
+We now offer the ability for developers to define their own substitutions as a JSON file and SerialNumberPattern will try to replace all matching input occurrences with the desired character(s).
+This is right now only available when developers have a skill that focuses on post-processing, and is using SerialNumberPattern and SerialNumberTextGroup.
+To make this work, be sure to include all the substitutions for a particular language in a file name like this "substitution-<language code>.json" in the same directory as appsettings.json.
+For example, for substitutions in English, please make a file with the name: substitution-en.json, and have the content as below:
+{
+  "substitutions": [
+    {
+      "substring": "DENIED",
+      "replace_with": "D9"
```
[the remark this answers is not in the document]
Comment (ruicliu) [reply experimental/SerialNumPostProcessingSkillSample/SpeechSerialNumber/SerialNumber.md]: Done, please see my latest commit.

File: experimental/SerialNumPostProcessingSkillSample/SpeechSerialNumber/SerialNumberPattern.cs (hunk 2 above)
Comment (ruicliu) [reply experimental/SerialNumPostProcessingSkillSample/SpeechSerialNumber/SerialNumberPattern.cs]: Moved this to InferMatch, please see my latest commit, and let me know if it's good!

**Review by tsuwandy — APPROVED** [review]

**Pull request merged**